drivers/adc/adc_mcux_lpadc.c

```diff
@@ -57,7 +59,119 @@ struct mcux_lpadc_data {
+{
+	const struct mcux_lpadc_config *config = dev->config;
+	uint32_t adc_freq_hz = 0;
+	float adc_cycle_ns = 0;
+	/* The unit of elements in the sample time array is ADCK. */
+	uint32_t acquisition_time_value = ADC_ACQ_TIME_VALUE(acq_time);
+	uint8_t acquisition_time_unit = ADC_ACQ_TIME_UNIT(acq_time);
+	uint32_t sample_time_array[] = {3, 5, 7, 11, 19, 35, 67, 131};
+	uint8_t sample_time_array_size = ARRAY_SIZE(sample_time_array);
+	uint8_t array_left_index = 0;
+	uint8_t array_right_index = sample_time_array_size - 1;
+	uint8_t array_mid_index = (array_left_index + array_right_index) / 2;
+	int left_gap = 0;
+	int right_gap = 0;
+
+	if (acquisition_time_value == ADC_ACQ_TIME_DEFAULT) {
+		return 0;
+	}
+
+	/* If the acquisition time is expressed in ADC ticks, direct match the option. */
+	if (ADC_ACQ_TIME_TICKS == acquisition_time_unit) {
+		switch (acquisition_time_value) {
+		case 3:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK3;
+			break;
+		case 5:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK5;
+			break;
+		case 7:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK7;
+			break;
+		case 11:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK11;
+			break;
+		case 19:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK19;
+			break;
+		case 35:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK35;
+			break;
+		case 67:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK67;
+			break;
+		case 131:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK131;
+			break;
+		default:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK7;
+			return -EINVAL;
+		}
+	}
+	/* If the acquisition time is expressed in nanoseconds or microseconds, each option in the
+	 * sample time array should be converted to nanoseconds, and then find the closest
+	 * configuration option.
+	 */
+	else {
+		/* Get ADC functional clock frequency to configure ADC acquisition time. */
+		if (clock_control_get_rate(config->clock_dev, config->clock_subsys, &adc_freq_hz)) {
+			LOG_ERR("Get device LPADC operating frequency failed");
+			return -EINVAL;
+		}
+
+		adc_cycle_ns = ((float)1000000000.0 / adc_freq_hz);
+
+		for (uint8_t index = 0; index < sample_time_array_size; ++index) {
+			sample_time_array[index] =
+				(uint32_t)(sample_time_array[index] * adc_cycle_ns);
+		}
+
+		if (ADC_ACQ_TIME_MICROSECONDS == acquisition_time_unit) {
+			acquisition_time_value *= 1000;
+		}
+		/* If the acquisition time that needs to be configured is less than the minimum
+		 * value that can be configured, configure the minimum value into the register. If
+		 * the acquisition time that needs to be configured is greater than the maximum
+		 * value that can be configured, configure the maximum value into the register.
+		 * Otherwise, compare the acquisition time that needs to be configured with the
+		 * midpoint of the array, and iterate the left or right boundary of the array based
+		 * on the comparison result until the most appropriate configuration option is
+		 * found.
```

Comment:
I don't think this is the correct approach. You should rather take the closest configurable value that is not shorter than the requested acquisition time. And return -EINVAL if that requested time exceeds what can be configured.

Reply:
> I don't think this is the correct approach. You should rather take the closest configurable value that is not shorter than the requested acquisition time. And return -EINVAL if that requested time exceeds what can be configured.

Thanks for your comments, I'd like to know why you think the configure value should be greater than the input acquisition time.

Reply:
The following statement explains why I use the floating point value. I still do not know why **configure value should be greater than the input acquisition time** just like @anangl said. Seems this code caused trouble to many people, maybe it is enough to only support the acquisition time being passed in ADCK units? 

Why I think the variable adc_cycle_ns needs to be set to a floating point type, I would like to give an example to illustrate. Under NXP LPC55S36 platform, the default ADC clock frequency is 48MHz, so the ADC cycle is approximately 20.833 ns (if variable adc_cycle_ns is set to an integer type, then this number is 20).

case1(float point type, 20.833):
sample_time_array[0] = 20.833 * 3 = 62.499 ns
sample_time_array[1] = 20.833 * 5 = 104.165 ns
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option kLPADC_SampleTimeADCK3 will be set to the corresponding bitfield.

case1(integer type, 20):
sample_time_array[0] = 20 * 3 = 60 ns
sample_time_array[1] = 20 * 5 = 100 ns
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option kLPADC_SampleTimeADCK5 will be set to the corresponding bitfield.

If we use integer type in the code, the accuracy of the calculation results decreases. At this clock frequency (48MHz), and application configuration options (82ns), the optimal configuration item should be kLPADC_SampleTimeADCK3.

Currently, my local code has changed the ADC sample time array (sample_time_array) to the integer type. The reason is that converting the final calculation result into an integer will not cause too much loss of accuracy, and at the same time reduces the calculation pressure of chips without FPU platforms (the subsequent calculation process uses array elements to compare with the acquisition_time_value).

for example
sample_time_array[0] = 20.833 * 3 = 62.499 ns = 62
sample_time_array[1] = 20.833 * 5 = 104.165 ns = 104
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option kLPADC_SampleTimeADCK3 will be set to the corresponding bitfield.